### PR TITLE
[AIR] Rename Keras callback and its parameters

### LINKED ISSUE #31212: [AIR] Rename `ray.air.integrations.keras.Callback` to be descriptive
### Description

Rename `Callback` as something descriptive like `MetricReporter` or `ModelCheckpoint`.

### Use case

A descriptive name would make my code more readable. Currently, the name tells me nothing about what the callback does (unlike TensorFlow's built-in callbacks).

```
model.fit(
        ...
        callbacks=[
            Callback(),
            LearningRateScheduler(...),
            EarlyStopping(),
        ], 
    )
```

### LINKED ISSUE #31213: [AIR] Checkpoint model and report metric at different steps
### Description

Update `ray.air.integrations.keras.Callback` to let users checkpoint and report metrics at different frequencies.

We could do this by splitting `Callback` into two distinct callbacks:

```
CheckpointModel(on="epoch_end)
ReportMetrics(on=("on_train_batch_end", "on_test_batch_end"), frequency=4)
```

### Use case

I want to checkpoint every epoch and report metrics to W&B every few batches. I can sort-of achieve this by using the `frequency` parameter, but this isn't ideal because:
1. I have to compute the correct `frequency`
2. I have the report metrics every batch (as opposed to every few)

```
Callback(on=("on_train_batch_end", "on_test_batch_end"), frequency=???)
```

### PR DESCRIPTION
**Pull request by bveeramani**
Signed-off-by: Balaji Veeramani <balaji@anyscale.com>

## Why are these changes needed?

The name `Callback` is undescriptive, and it's difficult to checkpoint and report metrics on different steps because you hae to manually compute `frequency`. This PR renames `ray.air.integrations.keras.Callback` as `ReportCheckpointCallback` and  replaces the `frequency` parameter with `checkpoint_on`. It also renames `on` as `report_metrics_on`.
 
## Related issue number

Closes #31212 and closes #31213 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/integrations/keras.py [hunk 1]
```diff
@@ -111,7 +102,89 @@ def on_predict_end(self, logs=None):
...
+        .. code-block: python
+
+            ############# Using it in TrainSession ###############
+            from ray.air.integrations.keras import Callback
+            def train_loop_per_worker():
+                strategy = tf.distribute.MultiWorkerMirroredStrategy()
+                with strategy.scope():
+                    model = build_model()
+                    #model.compile(...)
+                model.fit(dataset_shard, callbacks=[Callback()])
```
Comment (krfricke) [inline python/ray/air/integrations/keras.py:123]:
```suggestion
            ############# Using it in TrainSession ###############
            from ray.air.integrations.keras import ReportCheckpointCallback
            def train_loop_per_worker():
                strategy = tf.distribute.MultiWorkerMirroredStrategy()
                with strategy.scope():
                    model = build_model()

                model.fit(dataset_shard, callbacks=[ReportCheckpointCallback()])
```

File: python/ray/air/integrations/keras.py [hunk 2]
```diff
@@ -111,7 +102,89 @@ def on_predict_end(self, logs=None):
...
+
+    Args:
+        metrics: Metrics to report. If this is a list, each item describes
+            the metric key reported to Keras, and it's reported under the
+            same name. If this is a dict, each key is  the name reported
```
Comment (krfricke) [inline python/ray/air/integrations/keras.py:128]:
```suggestion
            same name. If this is a dict, each key is the name reported
```

File: python/ray/air/tests/test_keras_callback.py [hunk 3]
```diff
@@ -13,6 +17,114 @@
 )
 
 
+class TestReportCheckpointCallback:
```
Comment (krfricke) [inline python/ray/air/tests/test_keras_callback.py:20]: Why do we need the class here?

File: python/ray/air/tests/test_keras_callback.py [hunk 4]
```diff
@@ -13,6 +17,114 @@
...
+    @pytest.mark.parametrize(
+        "metrics, expected_metrics_keys",
+        [
+            (None, {"loss", "accuracy", "val_loss", "val_accuracy"}),
+            ("loss", {"loss"}),
+            (["loss", "accuracy"], {"loss", "accuracy"}),
+            ({"spam": "loss"}, {"spam"}),
+        ],
+    )
+    def test_metrics_keys(self, mock_report, metrics, expected_metrics_keys, model):
```
Comment (krfricke) [inline python/ray/air/tests/test_keras_callback.py:43]: Can we add short descriptions on what each test does and what we assert?

**Review by krfricke — COMMENTED** [review]
Looks good, only minor comments

File: python/ray/air/tests/test_keras_callback.py (hunk 3 above)
Comment (bveeramani) [reply python/ray/air/tests/test_keras_callback.py:20]: We don't. It was a stylistic choice to group the `ReportCheckpointCallback` tests together. Without it, we may need to include something like `report_checkpoint_callback` in the test names to distinguish them from the old tests.

Comment (krfricke) [reply python/ray/air/tests/test_keras_callback.py:20]: Ok that's fine with me

**Review by krfricke — APPROVED** [review]
Thanks!
